src/core/analytics/visibilityanalyticshandler.js

```diff
@@ -0,0 +1,60 @@
+import { isIE } from '../utils/useragent';
+import AnalyticsEvent from '../analytics/analyticsevent';
+import Searcher from '../models/searcher';
+
+/**
+ * Manages the document's visibility status and handles any visibility related analytics events.
+ */
+export default class VisibilityAnalyticsHandler {
+  constructor (analyticsReporterService, verticalKey) {
+    this._documentVisibilityState = undefined;
+    this._analyticsReporterService = analyticsReporterService;
+    this._verticalKey = verticalKey;
+  }
+
+  /**
+   * Initialize visibility change event listener(s) to send analytics events
+   * when a result page have become visible or have been hidden.
+   */
+  initVisibilityChangeListeners () {
+    /**
+     * Safari desktop listener and IE11 listeners fire visibility change event twice when switch
+     * to new tab and then close browser. This variable is used to ensure RESULTS_HIDDEN analytics event
+     * does not get send again if the page is already hidden.
+     */
+    document.addEventListener('visibilitychange', () => {
+      if (document.visibilityState === 'hidden' && this._documentVisibilityState !== 'hidden') {
+        this._documentVisibilityState = 'hidden';
+        this._reportVisibilityChangeEvent('RESULTS_HIDDEN');
+      } else if (document.visibilityState === 'visible') {
```

Comment:
do we need to check that `this._documentVisibilityState` is not `'visible'` here as well?

Reply:
I don't see any scenarios where it go from visible to visible state that it would double fire the analytics so I believe we don't need to check it. The reason it is needed for the hidden state is that safari desktop and unload event use by ie11 trigger this event when browser is close even through user have switch to a different tab so there would be a chance of it going from hidden to hidden state.